### DISCUSSION TITLE
MPI & web compatibility

### DISCUSSION THREAD

**Discussion by iakovdudinsky** [mpi4py/mpi4py#467] [Q&A]
I am developing a web application using HDF5 for data management and I need MPI for parallel data processing. However, I didn't find any case on using MPI with ASGI server. I start several workers with hypercorn, but if I launch `mpiexec -n 4 hypercorn ...` only one worker will work correctly due to binding.

Is MPI compatible with hypercorn/uvicorn/sanic or any ASGI server in general?

**Comment by dalcinl** [comment]
Being just an specification, MPI by itself is compatible with almost anything. However, specific implementations of MPI like MPICH or Open MPI may have runtime constraints that may not support every possible approach to server-side app architecture.

Disclaimer: I have no idea what hypercorn/unicorn/sanic are, web development it's not my line of work. However, I'm almost sure you SHOULD NOT `mpiexec -n 4 hypercorn ...`, that's never going to work. You should instead execute a single hypercorn process, and a use separate parallel MPI server to process HDF5 files. There are two possible approaches:

1. You start the MPI processes separately from your server-side web app. The MPI processes limit themselves to handle HDF5 fiels. Then you communicate MPI rank 0 with your hypercorn process via more traditional IPC mechanisms like sockets or pipes. 
2. Alternatively, you use MPI spawning (`MPI.Comm.Spawn`) in MPI. That way, you start your hypercorn process with `mpiexec -n 1 ...`, and you use Spawn to create new processes for the HDF5 heavy lifting. This way your server-side web app becomes a client for the parallel HDF5 server, and all client-server communication goes through MPI. If you go this route, I would suggest you use MPICH as the MPI implementation. Open MPI may work just fine, but it may not. Beware that MPI implementations are not interchangeable, so your mpi4py has to be built with the same MPI your HDF5 is built.
